### PR TITLE
Issue 2997 - Download original file endpoint

### LINKED ISSUE #2997: Endpoint to download the original files
### Description
v4 endpoint: https://3drepo.github.io/3drepo.io/#api-Model-downloadModel
proposed v5 endpoint: `GET /teamspaces/{teamspace}/projects/{project}/containers/{container}/revisions/{revision}/files`

This should return the original file (e.g. rvt, obj etc) as a binary stream.

NOTE: in v4 you can only download the original files from the latest revision, from v5 onwards we want to be able to download any revisions

### Goals
- create endpoint to allow users to download the model files

### Tasks
To be filled in by the developer

### PR DESCRIPTION
**Pull request by ChristosTsiotsias**
This fixes #2997

#### Description
Endpoint to download original files from a revision
Swagger doc, unit and e2e tests are included

#### Test cases
The user should be able to download the original file of a revision in binary form.
Swagger documentation should be present.

### CODE REVIEW

File: backend/tests/v5/e2e/routes/teamspaces/projects/containers/revisions.test.js [hunk 1]
```diff
@@ -88,14 +92,15 @@ const setupData = async () => {
...
+		ServiceHelper.db.createRevisionRef(teamspace, modelWithRev._id, `${validRefTypeRevision._id}P2006823-BDP-XX-A-Stanmore-Envelope_optimized_ifc`),
 	]);
 };
 
 const formatRevisions = (revs, includeVoid = false) => {
 	const formattedRevisions = revs
 		.sort((a, b) => b.timestamp - a.timestamp)
 		.flatMap((rev) => (includeVoid
-			|| !rev.void ? { ...rev, timestamp: rev.timestamp.getTime() } : []));
+			|| !rev.void ? { ...rev, timestamp: rev.timestamp.getTime(), rFile: undefined } : []));
```
Comment (carmenfan) [inline backend/tests/v5/e2e/routes/teamspaces/projects/containers/revisions.test.js:103]: rFile is an internal data and it should never be exposed to the outside world, even as an undefined - check what the code is doing if you're seeing this escaping to the outside world?

File: backend/tests/v5/e2e/routes/teamspaces/projects/containers/revisions.test.js [hunk 2]
```diff
@@ -296,6 +301,51 @@ const testNewRevision = () => {
...
+			const res = await agent.get(`${route(teamspace, project.id, modelWithRev._id, revisions[0]._id)}?key=${users.noProjectAccess.apiKey}`).expect(templates.notAuthorized.status);
+			expect(res.body.code).toEqual(templates.notAuthorized.code);
+		});
+
+		test('should fail if revision has no file', async () => {
+			const res = await agent.get(`${route(teamspace, project.id, modelWithRev._id, noFileRevision._id)}?key=${users.tsAdmin.apiKey}`).expect(templates.noFileFound.status);
+			expect(res.body.code).toEqual(templates.noFileFound.code);
+		});
+
+		test('should fail if revision has file but no ref', async () => {
```
Comment (carmenfan) [inline backend/tests/v5/e2e/routes/teamspaces/projects/containers/revisions.test.js:338]: check permission level too

Ensure only collaborator+ can get the file

File: backend/tests/v5/e2e/routes/teamspaces/projects/containers/revisions.test.js [hunk 3]
```diff
@@ -296,6 +301,51 @@ const testNewRevision = () => {
...
+			const res = await agent.get(`${route(teamspace, project.id, modelWithRev._id, noRefRevision._id)}?key=${users.tsAdmin.apiKey}`).expect(templates.noFileFound.status);
+			expect(res.body.code).toEqual(templates.noFileFound.code);
+		});
+
+		test('should download files if there is a valid ref', async () => {
+			await agent.get(`${route(teamspace, project.id, modelWithRev._id, validRefTypeRevision._id)}?key=${users.tsAdmin.apiKey}`).expect(templates.ok.status);
```
Comment (carmenfan) [inline backend/tests/v5/e2e/routes/teamspaces/projects/containers/revisions.test.js:344]: Should also be checking the content of the file is what you're expecting.

File: backend/tests/v5/e2e/routes/teamspaces/projects/containers/revisions.test.js [hunk 4]
```diff
@@ -88,14 +92,15 @@ const setupData = async () => {
 		ServiceHelper.db.createUser(nobody),
 		ServiceHelper.db.createProject(teamspace, project.id, project.name, models.map(({ _id }) => _id)),
 		...revisions.map((revision) => ServiceHelper.db.createRevision(teamspace, modelWithRev._id, revision)),
+		ServiceHelper.db.createRevisionRef(teamspace, modelWithRev._id, `${validRefTypeRevision._id}P2006823-BDP-XX-A-Stanmore-Envelope_optimized_ifc`),
```
Comment (carmenfan) [inline backend/tests/v5/e2e/routes/teamspaces/projects/containers/revisions.test.js:95]: let createRevision do this if there's a file ref, just pass data with it to store

File: backend/tests/v5/helper/services.js [hunk 5]
```diff
@@ -95,6 +96,13 @@ db.createRevision = (teamspace, modelId, revision) => {
 	return DbHandler.insertOne(teamspace, `${modelId}.history`, formattedRevision);
 };
 
+db.createRevisionRef = async (teamspace, modelId, refId) => {
+	const data = 'test data';
+	let refInfo = await ExternalServices.storeFile(teamspace, `${modelId}.history.ref`, data);
```
Comment (carmenfan) [inline backend/tests/v5/helper/services.js:101]: We need to make sure we clean the folder after the tests or we end up with clutter, or files that shouldn't exist to start existing.

the fileshare is set here https://github.com/3drepo/3drepo.io/blob/master/config/testV5/config.js#L82

So as part of the tear down (the bit that is called in [`afterAll` in jest](https://jestjs.io/docs/setup-teardown#one-time-setup)), you will probabyl want a function to remove all contents within that folder.

File: backend/tests/v5/helper/services.js [hunk 6]
```diff
@@ -164,12 +172,22 @@ ServiceHelper.generateUserCredentials = () => ({
...
-	void: !!isVoid,
+ServiceHelper.generateRevisionEntry = (isVoid = false, hasFile = true) => {
+	const _id = ServiceHelper.generateUUIDString();
+	return {
+		_id,
+		tag: ServiceHelper.generateRandomString(),
+		author: ServiceHelper.generateRandomString(),
+		timestamp: ServiceHelper.generateRandomDate(),
+		void: !!isVoid,
+		rFile: hasFile ? [`${_id}P2006823-BDP-XX-A-Stanmore-Envelope_optimized_ifc`] : undefined,
```
Comment (carmenfan) [inline backend/tests/v5/helper/services.js:183]: should only has this entry if it has the file - we shouldn't be storing `rFile: undefined` anywhere.
```suggestion
		...(hasFile ? {rFile: [`${_id}P2006823-BDP-XX-A-Stanmore-Envelope_optimized_ifc`]} : {})
```

Comment (carmenfan) [reply backend/tests/v5/helper/services.js:183]: why do we need to hardcode in `P2006823-BDP-XX-A-Stanmore-Envelope_optimized_ifc` ?

File: backend/tests/v5/helper/services.js [hunk 7]
```diff
@@ -164,12 +172,22 @@ ServiceHelper.generateUserCredentials = () => ({
...
+	};
+};
+
+ServiceHelper.generateRevisionRef = (revisionId, hasValidType = true) => ({
+	_id: `${revisionId}P2006823-BDP-XX-A-Stanmore-Envelope_optimized_ifc`,
```
Comment (carmenfan) [inline backend/tests/v5/helper/services.js:188]: hardcoded model name here too

File: backend/tests/v5/unit/models/fileRefs.test.js [hunk 8]
```diff
@@ -133,7 +145,65 @@ const testRemoveAllFilesFromModel = () => {
...
+			const res = await FileRefs.downloadFiles('someTS', 'someModel', revision);
+			expect(res).toEqual(expectedData);
+			expect(fn.mock.calls.length).toBe(1);
+			expect(fn.mock.calls[0][0]).toEqual('someTS');
+			expect(fn.mock.calls[0][1]).toEqual('someModel.history.ref');
+			expect(fn.mock.calls[0][2]).toEqual({ _id: filename });
+		});
+
+		test('should download the revision files', async () => {
+			const filename = 'b74ba13b-71db-4fcc-9ff8-7f640aa3dec2P2006823-BDP-XX-A-Stanmore-Envelope_optimized_ifc';
```
Comment (carmenfan) [inline backend/tests/v5/unit/models/fileRefs.test.js:176]: too ad-hoc, just use generateUUIDString() and generateRandomString() to make up this name.

File: backend/tests/v5/unit/models/fileRefs.test.js [hunk 9]
```diff
@@ -133,7 +145,65 @@ const testRemoveAllFilesFromModel = () => {
...
+			const res = await FileRefs.downloadFiles('someTS', 'someModel', revision);
+			expect(res).toEqual(expectedData);
+			expect(fn.mock.calls.length).toBe(1);
+			expect(fn.mock.calls[0][0]).toEqual('someTS');
+			expect(fn.mock.calls[0][1]).toEqual('someModel.history.ref');
+			expect(fn.mock.calls[0][2]).toEqual({ _id: filename });
+		});
+
+		test('should download the revision files with no valid extension', async () => {
+			const filename = 'b74ba13b-71db-4fcc-9ff8-7f640aa3dec2P2006823-BDP-XX-A-Stanmore-Envelope';
```
Comment (carmenfan) [inline backend/tests/v5/unit/models/fileRefs.test.js:190]: same here

File: backend/tests/v5/unit/models/fileRefs.test.js [hunk 10]
```diff
@@ -133,7 +145,65 @@ const testRemoveAllFilesFromModel = () => {
 	});
 };
 
+const testDownloadFiles = () => {
+	describe('Download files', () => {
+		test('should throw error if the revision has no files', async () => {
+			const revision = { rFile: [] };
```
Comment (carmenfan) [inline backend/tests/v5/unit/models/fileRefs.test.js:151]: also add a case for if rFile is not defined

**Review by carmenfan — CHANGES_REQUESTED** [review]

File: backend/src/v5/processors/teamspaces/projects/models/containers.js [hunk 11]
```diff
@@ -83,6 +84,11 @@ Containers.newRevision = (teamspace, model, data, file) => queueModelUpload(team
 
 Containers.updateRevisionStatus = updateRevisionStatus;
 
+Containers.downloadRevisionFiles = async (teamspace, container, revisionId) => {
+	const revision = await getRevisionById(teamspace, container, revisionId, { rFile: 1 });
+	return downloadFiles(teamspace, container, revision);
```
[the remark this answers is not in the document]
Comment (ChristosTsiotsias) [reply backend/src/v5/processors/teamspaces/projects/models/containers.js]: do you reckon I should add another method in fileRef model called getRevisionByTag or should I expose findOneRevisionByQuery and call that with a query from the processor ?

[the remark this answers is not in the document]
Comment (carmenfan) [reply backend/src/v5/processors/teamspaces/projects/models/containers.js]: you mean revision model? 

I think you should just have a getRevisionByIdOrTag. In general all our revision queries are based on that

**Review by carmenfan — APPROVED** [review]

**Review by carmenfan — APPROVED** [review]

**Pull request merged**